R/class-spec.R

```diff
@@ -0,0 +1,150 @@
+#' Standard class specifications
+#'
+#' This is used as the interface between R7 and R's other OO systems, allowing
+#' you to build R7 classes and methods on top of base types, informal S3
+#' classes, and formal S4 classes.
+#'
+#' @param x A class specification.
+#'    * An R7 class object or class union.
+#'    * An S3 class object, created by `s3_class()`.
+#'    * An S4 class object.
+#'    * A base type specified either with its constructor (`logical`, `integer`,
+#'      `double` etc) or its name (`"logical"`, `"integer"`, "`double`" etc).
+#' @param arg_name Argument name used when generating errors.
+#' @export
+as_class <- function(x, arg_name = "as_class()") {
+  if (is.null(x)) {
+    x
+  } else if (is_class(x)) {
+    x
+  } else if (is_union(x)) {
+    x
+  } else if (is_s3_class(x)) {
+    x
+  } else if (isS4(x) && methods::is(x, "classGeneratorFunction")) {
+    methods::getClass(as.character(x@className))
+  } else if (isS4(x) && methods::is(x, "classRepresentation")) {
+    x
+  } else if (is.function(x)) {
+    candidate <- Filter(function(y) identical(x, y), base_constructors)
+    if (length(candidate)  == 0) {
+      stop(sprintf("%s: could not find base class corresponding to supplied constructor function", arg_name), call. = FALSE)
+    }
+    base_classes[[names(candidate)[[1]]]]
+  } else if (is.character(x) && length(x) == 1) {
+    if (x %in% names(base_classes)) {
+      base_classes[[x]]
+    } else {
+      stop(sprintf("%s: Can't find base class called '%s'", arg_name, x), call. = FALSE)
+    }
+  } else {
+    stop(sprintf("%s: class specification must by a R7 class object, result of s3_class(), a S4 class object, or a base constructor function, not a %s.", arg_name, obj_desc(x)), call. = FALSE)
```

Comment:
```suggestion
    stop(sprintf("%s: class specification must be an R7 class object, the result of `s3_class()`, an S4 class object, or a base constructor function, not a %s.", arg_name, obj_desc(x)), call. = FALSE)
```

+ regenerate snapshots